source/mqtt311_decoder.c

```diff
@@ -0,0 +1,206 @@
+     * states if the current MQTT packet was received in a fragmented manner.
+     */
+    struct aws_byte_cursor temp_cursor = *data;
+    struct aws_mqtt_fixed_header packet_header;
+    AWS_ZERO_STRUCT(packet_header);
+    if (!aws_mqtt_fixed_header_decode(&temp_cursor, &packet_header) &&
+        temp_cursor.len >= packet_header.remaining_length) {
+
+        /* figure out the cursor that spans the full packet */
+        size_t fixed_header_length = temp_cursor.ptr - data->ptr;
+        struct aws_byte_cursor whole_packet_cursor = *data;
+        whole_packet_cursor.len = fixed_header_length + packet_header.remaining_length;
+
+        /* advance the external, mutable data cursor to the start of the next packet */
+        aws_byte_cursor_advance(data, whole_packet_cursor.len);
+
+        /*
+         * if this fails, the decoder goes into an error state.  If it succeeds we'll loop again into the same state
+         * because we'll be back at the beginning of the next packet (if it exists).
+         */
+        enum aws_mqtt_packet_type packet_type = aws_mqtt_get_packet_type(whole_packet_cursor.ptr);
+        return aws_result_to_mqtt311_decoding_directive(
+            s_aws_mqtt311_decoder_safe_packet_handle(decoder, packet_type, whole_packet_cursor));
+    }
+
+    uint8_t byte = *data->ptr;
+    aws_byte_cursor_advance(data, 1);
+    aws_byte_buf_append_byte_dynamic(&decoder->packet_buffer, byte);
+
+    decoder->state = AWS_MDST_READ_REMAINING_LENGTH;
+
+    return AWS_MDD_CONTINUE;
+}
+
+static enum aws_mqtt311_decoding_directive s_handle_decoder_read_remaining_length(
+    struct aws_mqtt311_decoder *decoder,
+    struct aws_byte_cursor *data) {
+    AWS_FATAL_ASSERT(decoder->total_packet_length == 0);
+    if (data->len == 0) {
+        return AWS_MDD_OUT_OF_DATA;
+    }
+
+    uint8_t byte = *data->ptr;
+    aws_byte_cursor_advance(data, 1);
+    aws_byte_buf_append_byte_dynamic(&decoder->packet_buffer, byte);
+
+    struct aws_byte_cursor vli_cursor = aws_byte_cursor_from_buf(&decoder->packet_buffer);
+    aws_byte_cursor_advance(&vli_cursor, 1);
+
+    size_t remaining_length = 0;
+    int vli_result = aws_mqtt311_decode_remaining_length(&vli_cursor, &remaining_length);
+
+    /* anything other than a short buffer error (not enough data yet) is a terminal error */
+    if (vli_result == AWS_OP_ERR && aws_last_error() != AWS_ERROR_SHORT_BUFFER) {
+        return AWS_MDD_PROTOCOL_ERROR;
+    }
+
+    /*
+     * If we successfully decoded a variable-length integer, we now know exactly how many bytes we need to receive in
+     * order to have the full packet.
+     */
+    if (vli_result == AWS_OP_SUCCESS) {
+        decoder->total_packet_length = remaining_length + decoder->packet_buffer.len;
+        AWS_FATAL_ASSERT(decoder->total_packet_length > 0);
+        decoder->state = AWS_MDST_READ_BODY;
+    }
+
+    return AWS_MDD_CONTINUE;
```

Comment:
trivial/style: I'd find the logic easier to follow if it were a bunch of if/elses. vs if-unexpected-error, if-success, implicitly-else this is an errror we were expecting
```suggestion
    if (aws_mqtt311_decode_remaining_length(&vli_cursor, &remaining_length) == AWS_OP_ERR) {
        if (aws_last_error() == AWS_ERROR_SHORT_BUFFER) {
            /* not enough data yet */
            return AWS_MDD_CONTINUE;
        } else {
            /* anything else is a terminal error */
            return AWS_MDD_PROTOCOL_ERROR;
        }
    } else {
        /* We successfully decoded a variable-length integer, we now know
         * exactly how many bytes we need to receive in order to have the full packet. */
        decoder->total_packet_length = remaining_length + decoder->packet_buffer.len;
        AWS_FATAL_ASSERT(decoder->total_packet_length > 0);
        decoder->state = AWS_MDST_READ_BODY;
    }
```

Reply:
switched, but dropped the outer else